lib/SILOptimizer/FunctionSignatureTransforms/ArgumentExplosionTransform.cpp

```diff
@@ -26,42 +35,216 @@ static llvm::cl::opt<bool> FSODisableArgExplosion(
+  unsigned const liveLeafCountUpperBound = argDesc.ProjTree.getLiveLeafCount();
+
+  assert(liveLeafCountUpperBound >= 0 && "Counting live leaves should result in a non-negative number.");
+
+  // If the argument is completely dead, it will not be exploded.
+  //
+  // Explosion makes sense only if some but not all of the leaves are live.  The
+  // dead argument transformation will try to eliminate the argument.
+  if (liveLeafCountUpperBound == 0) {
+    LLVM_DEBUG(llvm::dbgs() << "The argument has no live leaves.");
+    return false;
+  }
+
+  // To determine whether some but not all of the leaves are used, the total 
+  // leaf count must be retrieved.  
+  llvm::SmallVector<SILType, 32> allLeaves;
+  argDesc.ProjTree.getAllLeafTypes(allLeaves);
+  unsigned leafCount = allLeaves.size();
+
+  assert(liveLeafCountUpperBound <= leafCount && "There should be no more *live* leaves than there are *total* leaves.");
+
+  // NOTE: Exiting early at this point in the case where liveLeafCountUpperBound ==
+  //       leafCount misses cases where leaves are only live until the
+  //       owned-to-guaranteed transform eliminates their liveness.
+
+  unsigned const nontrivialLeafCount =
+      llvm::count_if(allLeaves, [&](SILType type) { return !type.isTrivial(*function); });
+
+  assert(nontrivialLeafCount >= 0 && "Counting non-trivial leaves should result in a non-negative number.");
+
+  // If the argument is completely trivial, it will not be exploded.
+  if (nontrivialLeafCount == 0) {
+    llvm_unreachable("There should be at least one nontrivial leaf since our type is not trivial.");
+    return false;
+  }
+
+  llvm::SmallVector<const ProjectionTreeNode *, 32> liveLeaves;
+  argDesc.ProjTree.getLiveLeafNodes(liveLeaves);
+  // NOTE: The value obtained here is an upper bound because the 
+  //       owned-to-guaranteed transformation may eliminate some live leaves,
+  //       leaving the count lower.
+  unsigned liveNontrivialLeafCountUpperBound =
+      llvm::count_if(liveLeaves, [&](const ProjectionTreeNode *leaf) {
+        return !leaf->getType().isTrivial(*function);
+      });
+
+  assert(liveNontrivialLeafCountUpperBound >= 0 && "Counting live non-trivial leaves should result in a non-negative number.");
+
+  assert(liveNontrivialLeafCountUpperBound <= nontrivialLeafCount && "There should be no more *live* non-trivial leaves than there are *total* non-trivial leaves.");
+  assert(nontrivialLeafCount <= leafCount && "There should be no more *non-trivial* leaves than there are *total* leaves.");
+
+  // The heuristic max explosion size for a single argument.
+  unsigned const maxExplosionSize = 3;
+
+  // If it is known without taking the owned-to-guaranteed transformation into
+  // account both that exploding will reduce ARC traffic (because an upper bound
+  // for the number of live non-trivial leaves is less thann the non-trivial 
+  // leaf count) and also that the explosion will fit within the heuristic upper
```

Comment:
Actually, when I think about it: is this really true, given that we pass most of the arguments as guaranteed by default? I think it mainly depends on the call site. Only if we need a retain-release pair at the call site, eliminating a dead argument will give a benefit.

Reply:
@eeckstein it is not that simple. What if we want to move that value pass the call site. Consider a dead guaranteed parameter. Without eliminating the argument we will not be able to reduce the lifetime and do other ARC optimizations.

I would put it this way, it only makes sense if we eliminate an ARC "constraint" on the lifetime of the non-trivially typed thing.